### DISCUSSION TITLE
ValueError: {'code': -32000, 'message': 'invalid sender'}

### DISCUSSION THREAD

**Discussion by whlustb** [Elnaril/uniswap-universal-router-decoder#45] [Q&A] [answered]
hello, can I get any help from here?  I used the your example in sell_token.py, but it keep report this error below.

ValueError: {'code': -32000, 'message': 'invalid sender'}

**Comment by Elnaril** [comment]
Hello @whlustb 

Are you sure this error is raised by the code you mentioned? This error (`ValueError: {'code': -32000, 'message': 'invalid sender'}`) comes usually from another part of the code. You should check how you build and sign your transaction. For example, are you sure of the `chainId` you're using? What about the other transaction fields? What about the account/private key you use to sign the transaction?

I confirm your first code snippet is correct (though it also depends on the variable values), while the second one is not (should be payer_is_sender=False)

Hope it helps!

**Reply by whlustb** [reply]
thank you for your help, my first problem is caused by a wrong chain_id, and I can do go through the code now, but the problem is that, I can can't see the bought usdc/uni or other token in my wallet, and when I check the transaction in explore,   I can see there are only one Transaction Action , but as you can see there are 2 action in the call, wrap_eth and v2_swap_exact_in

https://sepolia.etherscan.io/tx/0x1c1f9cf7725dc6982c2fb860ffec7591fafd78f6797340ee7e9d34baa46c53cc


![1711024177964](https://github.com/Elnaril/uniswap-universal-router-decoder/assets/41858356/1fa8bbe5-1c5d-4409-b472-35f5670cecbf)


from uniswap_universal_router_decoder import FunctionRecipient, RouterCodec
from web3 import Account, Web3

private_key = "**********************"
chain_id = 11155111
rpc_endpoint = "https://eth-sepolia.g.alchemy.com/v2/****************"
uni_address = Web3.to_checksum_address('0x2f6f07cdcf3588944bf4c42ac74ff24bf56e7590') #usdc
uni_abi = '[{"inputs":[{"internalTyp*************'

amount_in = int(0.01 * 10**18)
min_amount_out = 365 * 10**18
weth_address = Web3.to_checksum_address('0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2')
path = [weth_address, uni_address]
ur_address = Web3.to_checksum_address("0xEf1c6E67703c7BD7107eed8303Fbe6EC2554BF6B")
ur_abi = '[{"inpu*******************'

codec = RouterCodec()
w3 = Web3(Web3.HTTPProvider(rpc_endpoint))
account = Account.from_key(private_key)
encoded_input = (
        codec
        .encode
        .chain()
        .wrap_eth(FunctionRecipient.ROUTER, amount_in)
        .v2_swap_exact_in(FunctionRecipient.SENDER, amount_in, min_amount_out, path, account.address, payer_is_sender=True)
        .build(codec.get_default_deadline())
)

trx_params = {
        "from": account.address,
        "to": ur_address,
        "gas": 500000,
        "maxPriorityFeePerGas": w3.eth.max_priority_fee,
        "maxFeePerGas": 50 * 10**9,
        "type": '0x2',
        "chainId": chain_id,
        "value": amount_in,
        "nonce": w3.eth.get_transaction_count(account.address),
        "data": encoded_input,
}

raw_transaction = w3.eth.account.sign_transaction(trx_params, account.key).rawTransaction
trx_hash = w3.eth.send_raw_transaction(raw_transaction)
print(f"Trx Hash: {trx_hash}")
receipt = w3.eth.wait_for_transaction_receipt(trx_hash, timeout=900)

uni_contract = w3.eth.contract(address=uni_address, abi=uni_abi)
uni_balance = uni_contract.functions.balanceOf(account.address).call()

**Reply by Elnaril** [reply]
@whlustb , I can see 2 errors:
 - The UR address on Sepolia is not `0xEf1c6E67703c7BD7107eed8303Fbe6EC2554BF6B` (which is on mainnet, no contract on Sepolia at this address) but `0x3fC91A3afd70395Cd496C647d5a6CC9D4B2b7FAD`, see [UR repo](https://github.com/Uniswap/universal-router/blob/main/deploy-addresses/sepolia.json)
 - `.v2_swap_exact_in(FunctionRecipient.SENDER, amount_in, min_amount_out, path, account.address, payer_is_sender=True)` is wrong (see my previous comment). That should be either `.v2_swap_exact_in(FunctionRecipient.SENDER, amount_in, min_amount_out, path, payer_is_sender=False)` or `.v2_swap_exact_in_from_balance(FunctionRecipient.SENDER, amount_out_min, path)`

The second point did not make the transaction to fail because of the first one.

**Reply by whlustb** [reply]
thank you so much, I have changed the universal rooter contract code and the v2_swap_eact_in as you suggested, but I still keep getting this error, as below {'code': -32000, 'message': 'replacement transaction underpriced'}


from uniswap_universal_router_decoder import FunctionRecipient, RouterCodec
from web3 import Account, Web3


private_key = "***********************"
chain_id = 11155111
rpc_endpoint = "https://eth-sepolia.g.alchemy.com/v2/*********************"

uni_address = Web3.to_checksum_address('0x1f9840a85d5af5bf1d1762f925bdaddc4201f984')
uni_abi = '[{"inputs"**********'

amount_in = int(0.1 * 10**18)
min_amount_out = 0
weth_address = Web3.to_checksum_address('0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2')
path = [weth_address, uni_address]
ur_address = Web3.to_checksum_address("0x3fC91A3afd70395Cd496C647d5a6CC9D4B2b7FAD")
ur_abi = '[{"inputs":[{"components":[{"*****************'

codec = RouterCodec()
w3 = Web3(Web3.HTTPProvider(rpc_endpoint))
account = Account.from_key(private_key)
encoded_input = (
        codec
        .encode
        .chain()
        .wrap_eth(FunctionRecipient.ROUTER, amount_in)
        .v2_swap_exact_in(FunctionRecipient.SENDER, amount_in, min_amount_out, path, payer_is_sender=False)
        .build(codec.get_default_deadline())
)

trx_params = {
        "from": account.address,
        "to": ur_address,
        "gas": 500000,
        "maxPriorityFeePerGas": w3.eth.max_priority_fee,
        "maxFeePerGas": 50 * 10**9,
        "type": '0x2',
        "chainId": chain_id,
        "value": amount_in,
        "nonce": w3.eth.get_transaction_count(account.address),
        "data": encoded_input,
}

raw_transaction = w3.eth.account.sign_transaction(trx_params, account.key).rawTransaction
trx_hash = w3.eth.send_raw_transaction(raw_transaction)
print(f"Trx Hash: {trx_hash}")
receipt = w3.eth.wait_for_transaction_receipt(trx_hash, timeout=900)

**Reply by Elnaril** [reply]
`{'code': -32000, 'message': 'replacement transaction underpriced'}` means you have already a pending transaction with the same nonce. To replace the pending trx with this one, increase the gas fees substantially. It uses to be by at least +12.5%. Not sure how much it is nowadays or on Sepolia ...

**Reply by whlustb** [reply]
thank u for your reply, I have wait for more than 1 hour, and have follow your suggest, changed the gas from 200000 to 1000000, but the error is still here, and I can't find the pending trasaction in the etherscan.io, I wonder if I can manually cancel that pending trasaction, or any other solution?

trx_params = {
        "from": account.address,
        "to": ur_address,
        "gas": 1000000,
        "maxPriorityFeePerGas": w3.eth.max_priority_fee,
        "maxFeePerGas": 50 * 10**9,
        "type": '0x2',
        "chainId": chain_id,
        "value": amount_in,
        "nonce": w3.eth.get_transaction_count(account.address),
        "data": encoded_input,
}

**Reply by whlustb** [reply]
I can get a trx hash if I add +3 in the w3.eth.get_transaction_count(account.address), but cant't get the transaction result, and can't find the etherscan.io

**Reply by Elnaril** [reply]
Here we're talking of `maxPriorityFeePerGas` (and to some extent `maxFeePerGas`), not `gas` which is just the maximum quantity of gas you're ready to use.

BTW, the gas values I used in my examples and tests are by no mean relevant to any other context. Determining the right amounts are left to the lib users. The reason why your transaction is still pending is likely because of the values you used.

**Reply by Elnaril** [reply]
yes, if you increase the nonce, you can send other transactions. But still, the pending transaction must be mined (or dropped by miners from the mempool, but in this case the transactions having a nonce too big will be rejected)

**Reply by Elnaril** [reply]
There is no weth contract @ `0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2` on Sepolia ... Look instead here: `0xfFf9976782d46CC05630D1f6eBAb18b2324d6B14`

You should check every line and should not take for granted any of them ...

Please note that your questions are not specific to this lib but to how Ethereum and related technologies work and would be better asked to their own community/support ... Hope it makes sense ...

Thank you for your interest in this lib! :)

**Reply by Elnaril** [reply, marked as answer]
To summarize, and to answer the initial question (`ValueError: {'code': -32000, 'message': 'invalid sender'}`), the issue was caused by a wrong `chainId` in the transaction parameters.

**Comment by whlustb** [comment]


**Reply by Elnaril** [reply]
> thank u so much, I will buy 2 cup of coffe for your kindlly help

Thank you, that would be very much appreciated! :)

For you last error:
 - Are you sure there is a V2 pool for your pair on Sepolia? (and the V2 protocol all together?)
 - There are v3 pools for fees 500, 3000, 10000 for your pair. I didn't check their liquidity though. Try the 3000 one with `.v3_swap_exact_in`

**Reply by Elnaril** [reply]
>  it finally trade successfully with v3_swap

Great! :)

> can you give me a usdt address of your wallet

You'll find it in the [wiki](https://github.com/Elnaril/uniswap-universal-router-decoder/wiki)'s footer. :)

**Reply by Elnaril** [reply]
Thank you @whlustb ! It is very much appreciated! :)
